### PR TITLE
dotnet: fix some problems with nuget-to-nix and make it deterministic

### PR DESCRIPTION
**Pull request by corngood**
###### Description of changes

This series of changes makes nuget-to-nix deterministic.  I've tested it using this script, which should cover everything in nixpkgs:

```
#!/bin/sh -e
$(nix-build -A github-runner.createDepsFile --no-out-link)/bin/create-deps-file pkgs/development/tools/continuous-integration/github-runner/deps.nix
$(nix-build -A ryujinx.fetch-deps --no-out-link) pkgs/applications/emulators/ryujinx/deps.nix
$(nix-build -A pinta.fetch-deps --no-out-link) pkgs/applications/graphics/pinta/deps.nix
$(nix-build -A ArchiSteamFarm.fetch-deps --no-out-link) pkgs/applications/misc/ArchiSteamFarm/deps.nix
$(nix-build -A python-language-server.fetch-deps --no-out-link) pkgs/development/dotnet-modules/python-language-server/deps.nix
$(nix-build -A omnisharp-roslyn.fetch-deps --no-out-link) pkgs/development/tools/omnisharp-roslyn/deps.nix
$(nix-build -A osu-lazer.fetch-deps --no-out-link) pkgs/games/osu-lazer/deps.nix
$(nix-build -A xivlauncher.fetch-deps --no-out-link) pkgs/games/xivlauncher/deps.nix
$(nix-build -A jellyfin.fetch-deps --no-out-link) pkgs/servers/jellyfin/nuget-deps.nix
$(nix-build -A eventstore.fetch-deps --no-out-link) pkgs/servers/nosql/eventstore/deps.nix
$(nix-build -A baget.fetch-deps --no-out-link) pkgs/servers/web-apps/baget/deps.nix
$(nix-build -A opentabletdriver.fetch-deps --no-out-link) pkgs/tools/X11/opentabletdriver/deps.nix
$(nix-build -A discordchatexporter-cli.fetch-deps --no-out-link) pkgs/tools/backup/discordchatexporter-cli/deps.nix
$(nix-build -A depotdownloader.fetch-deps --no-out-link) pkgs/tools/misc/depotdownloader/deps.nix
```

(TODO: find a way to run these automatically, possible by implementing --deps-only across all updaters?)

This is an example of the kind of non-deterministic change that would happen:

```
diff --git a/pkgs/development/tools/omnisharp-roslyn/deps.nix b/pkgs/development/tools/omnisharp-roslyn/deps.nix
index d3d626aff5e..6bb3acce3b3 100644
--- a/pkgs/development/tools/omnisharp-roslyn/deps.nix
+++ b/pkgs/development/tools/omnisharp-roslyn/deps.nix
@@ -82,7 +82,7 @@
   (fetchNuGet { pname = "Microsoft.SourceLink.Common"; version = "1.0.0"; sha256 = "1zxkpx01zdv17c39iiy8fx25ran89n14qwddh1f140v1s4dn8z9c"; })
   (fetchNuGet { pname = "Microsoft.SourceLink.GitHub"; version = "1.0.0"; sha256 = "029ixyaqn48cjza87m5qf0g1ynyhlm6irgbx1n09src9g666yhpd"; })
   (fetchNuGet { pname = "Microsoft.TestPlatform.ObjectModel"; version = "17.2.0"; sha256 = "0l05smcgjzdfa5f60f9q5lylap3i21aswxbava92s19bgv46w2rv"; })
-  (fetchNuGet { pname = "Microsoft.TestPlatform.TranslationLayer"; version = "17.2.0"; sha256 = "0yzwsmyb1pz761rg03zjimbqhngqj2vgzppp0ypqhdxbp5gmh2k9"; })
+  (fetchNuGet { pname = "Microsoft.TestPlatform.TranslationLayer"; version = "17.2.0"; sha256 = "10a8dh2bl9kgls5n7fqlg3c84qb2wxbabdw0c3in8vv5369bpaq9"; url = "https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v3/flat2/microsoft.testplatform.translationlayer/17.2.0/microsoft.testplatform.translationlayer.17.2.0.nupkg"; })
   (fetchNuGet { pname = "Microsoft.VisualStudio.SDK.EmbedInteropTypes"; version = "15.0.12"; sha256 = "083pva0a0xxvqqrjv75if25wr3rq034wgjhbax74zhzdb665nzsw"; })
   (fetchNuGet { pname = "Microsoft.VisualStudio.Setup.Configuration.Interop"; version = "1.14.114"; sha256 = "062mqkmjf4k6zm3wi9ih0lzypfsnv82lgh88r35fj66akihn86gv"; })
   (fetchNuGet { pname = "Microsoft.VisualStudio.Setup.Configuration.Interop"; version = "1.16.30"; sha256 = "14022lx03vdcqlvbbdmbsxg5pqfx1rfq2jywxlyaz9v68cvsb0g4"; })
@@ -120,11 +120,11 @@
   (fetchNuGet { pname = "NuGet.Protocol"; version = "6.3.0-preview.1.32"; sha256 = "1k1vjcgdjrpxw52qgmmcviiic72jsk9afblxcsr391rnz871jkbs"; url = "https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v3/flat2/nuget.protocol/6.3.0-preview.1.32/nuget.protocol.6.3.0-preview.1.32.nupkg"; })
   (fetchNuGet { pname = "NuGet.Versioning"; version = "5.2.0"; sha256 = "08ay8bhddj9yiq6h9lk814l65fpx5gh1iprkl7pcp78g57a6k45k"; })
   (fetchNuGet { pname = "NuGet.Versioning"; version = "6.3.0-preview.1.32"; sha256 = "1aamlp2k26xdrgq5sjyhw5nfxzhbrv87q2a96hh37ipxya11smgv"; url = "https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v3/flat2/nuget.versioning/6.3.0-preview.1.32/nuget.versioning.6.3.0-preview.1.32.nupkg"; })
-  (fetchNuGet { pname = "OmniSharp.Extensions.JsonRpc"; version = "0.19.0"; sha256 = "0m9lw21iz90ayl35f24ir3vbiydf4sjqw590qqgwknykpzsi1ai2"; })
-  (fetchNuGet { pname = "OmniSharp.Extensions.JsonRpc.Generators"; version = "0.19.0"; sha256 = "17akjdh9dnyxr01lnlsa41ca52psqnny8j3wxz904zs15pz932ln"; })
-  (fetchNuGet { pname = "OmniSharp.Extensions.LanguageProtocol"; version = "0.19.0"; sha256 = "06d4wakdaj42c9qnlhdyqrjnm97azp4hrvfg70f96ldl765y9vrf"; })
-  (fetchNuGet { pname = "OmniSharp.Extensions.LanguageServer"; version = "0.19.0"; sha256 = "0k1z3zchl1d82fj0ha63i54g5j046iaz8vb3cyxpjb6kp7zah28v"; })
-  (fetchNuGet { pname = "OmniSharp.Extensions.LanguageServer.Shared"; version = "0.19.0"; sha256 = "0s3h9v5p043ip27g9jcvd0np9q3hn2pfv6gn539m45yb5d74a6i5"; })
+  (fetchNuGet { pname = "OmniSharp.Extensions.JsonRpc"; version = "0.19.0"; sha256 = "1vxsf75bda1v20bjgyjd9n2jp1k6zdfnxqbg458iqrx7pc7078sl"; url = "https://az774065.vo.msecnd.net/artifacts/omnisharp/nuget/v3/flatcontainer/omnisharp.extensions.jsonrpc/0.19.0/omnisharp.extensions.jsonrpc.0.19.0.nupkg"; })
+  (fetchNuGet { pname = "OmniSharp.Extensions.JsonRpc.Generators"; version = "0.19.0"; sha256 = "1ll7yq4655d1i31yqv577k0lzsrmpizwpw3sjcwzw2sbzfh0c2rv"; url = "https://az774065.vo.msecnd.net/artifacts/omnisharp/nuget/v3/flatcontainer/omnisharp.extensions.jsonrpc.generators/0.19.0/omnisharp.extensions.jsonrpc.generators.0.19.0.nupkg"; })
+  (fetchNuGet { pname = "OmniSharp.Extensions.LanguageProtocol"; version = "0.19.0"; sha256 = "1p78aawai6mlw2k38m709bw87j3q2indjs8iiapai3mb7pm12lm8"; url = "https://az774065.vo.msecnd.net/artifacts/omnisharp/nuget/v3/flatcontainer/omnisharp.extensions.languageprotocol/0.19.0/omnisharp.extensions.languageprotocol.0.19.0.nupkg"; })
+  (fetchNuGet { pname = "OmniSharp.Extensions.LanguageServer"; version = "0.19.0"; sha256 = "043790kfzk05n90hhpmq2kxs2hdalm3rjq7smy80ap8lgsjw1sp0"; url = "https://az774065.vo.msecnd.net/artifacts/omnisharp/nuget/v3/flatcontainer/omnisharp.extensions.languageserver/0.19.0/omnisharp.extensions.languageserver.0.19.0.nupkg"; })
+  (fetchNuGet { pname = "OmniSharp.Extensions.LanguageServer.Shared"; version = "0.19.0"; sha256 = "0z7h7219crwdpfgbs6y2w596rs5p14p0kf5gm4mkj8m9hlg5kyij"; url = "https://az774065.vo.msecnd.net/artifacts/omnisharp/nuget/v3/flatcontainer/omnisharp.extensions.languageserver.shared/0.19.0/omnisharp.extensions.languageserver.shared.0.19.0.nupkg"; })
   (fetchNuGet { pname = "runtime.any.System.Collections"; version = "4.3.0"; sha256 = "0bv5qgm6vr47ynxqbnkc7i797fdi8gbjjxii173syrx14nmrkwg0"; })
   (fetchNuGet { pname = "runtime.any.System.Diagnostics.Tools"; version = "4.3.0"; sha256 = "1wl76vk12zhdh66vmagni66h5xbhgqq7zkdpgw21jhxhvlbcl8pk"; })
   (fetchNuGet { pname = "runtime.any.System.Diagnostics.Tracing"; version = "4.3.0"; sha256 = "00j6nv2xgmd3bi347k00m7wr542wjlig53rmj28pmw7ddcn97jbn"; })
```

I also fixed a couple of problems that were blocking the fetch scripts on some packages:

- intersect dotnet-sdk platforms with package platforms (otherwise a package could specify 'linux' and end up with unsupported platforms)
- roll forward a dependency in baget, which was causing downgrade errors

@IvarWithoutBones 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by corngood** [comment]
Also worth mentioning that I'm working on a fetcher that can canonicalise packages from different sources, but I'm not quite sure how it'll fit in yet.

**Comment by corngood** [comment]
@ofborg build ArchiSteamFarm baget depotdownloader eventstore github-runner jellyfin nuget-to-nix omnisharp-roslyn opentabletdriver osu-lazer pinta python-language-server ryujinx xivlauncher

**Comment by corngood** [comment]
I've locally built everything depending on this with the updated lock files, and confirmed that the lock files can all be updated without changes.

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh [hunk 1]
```diff
@@ -5,42 +5,77 @@ set -euo pipefail
...
+      if [[ "$source" == "$used_source" ]]; then
+        sha256="$(nix-hash --type sha256 --flat --base32 "$version/$package.$version".nupkg)"
+        found=true
+        break
+      else
+        if sha256=$(nix-prefetch-url "$url" 2>"$tmp"/error); then
+          echo $package $version is available on $url, but was downloaded from ${base_addresses[$used_source]}$package/$version/$package.$version.nupkg 1>&2
+          # exit 1
+          found=true
+          break
```
Comment (IvarWithoutBones) [inline pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:55]: I dont understand why this would happen, shouldn't the `url` always be fetched from? Now earlier endpoints in the `sources` array always take priority over later ones.

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh [hunk 2]
```diff
@@ -5,42 +5,77 @@ set -euo pipefail
...
+
+declare -A base_addresses
 
-while read pkg_spec; do
-  { read pkg_name; read pkg_version; } < <(
-    # Build version part should be ignored: `3.0.0-beta2.20059.3+77df2220` -> `3.0.0-beta2.20059.3`
-    sed -nE 's/.*<id>([^<]*).*/\1/p; s/.*<version>([^<+]*).*/\1/p' "$pkg_spec")
+for index in "${sources[@]}"; do
+  base_addresses[$index]=$(
+    curl --compressed -fsL "$index" | \
```
Comment (IvarWithoutBones) [inline pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:26]:
```suggestion
    curl --compressed --netrc -fsL "$index" | \
```

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh [hunk 3]
```diff
@@ -5,42 +5,77 @@ set -euo pipefail
...
+        else
+          if ! grep 'HTTP error 404' "$tmp/error" > /dev/null; then
+            cat "$tmp/error" 1>&2
+            exit 1
+          fi
+        fi
+      fi
+    done
+
+    if ! $found; then
```
Comment (IvarWithoutBones) [inline pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:65]: Just to avoid explicitly having to set a default above
```suggestion
    if ! ${found:-false}; then
```

**Review by IvarWithoutBones — COMMENTED** [review]
Thanks for the PR! This seems very useful, just a few questions/suggestions on the `nuget-to-nix` front. Could you rebase this as well?

**Comment by IvarWithoutBones** [comment]
This also doesn't seem to be sorted at all, have you checked if its deterministic across platforms? Not sure if the `for package in *` is consistent there.

**Comment by corngood** [comment]
https://www.gnu.org/software/bash/manual/bash.html#Filename-Expansion
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_13_03

It's supposed to be ordered, however, I think we'll need to set LC_COLLATE (or LC_ALL) for it to be safe.  I can make that change, or switch it back to using `sort`.  Whatever you prefer.

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh [hunk 4]
```diff
@@ -5,42 +5,77 @@ set -euo pipefail
...
-    continue
-  fi
+echo "{ fetchNuGet }: ["
 
-  pkg_sha256="$(nix-hash --type sha256 --flat --base32 "$(dirname "$pkg_spec")"/*.nupkg)"
+cd "$1"
+for package in *; do
+  cd "$package"
+  for version in *; do
+    id=$(xml sel -t -m //_:package/_:metadata/_:id -v . "$version/$package".nuspec)
```
[the remark this answers is not in the document]
Comment (corngood) [reply pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:36]: Cool.  I wasn't sure how accepted yq would be, but xmlstarlet is a pain in the ass, so I'll happily switch it over.

**Comment by IvarWithoutBones** [comment]
`LC_ALL` has already been used in combination with `sort` since [this commit](https://github.com/NixOS/nixpkgs/pull/162187/commits/bbcb7d3d34ca5ea620740184eea9db61e4b84f8b#diff-895a38210519585e536780762daa447ce10771e91be5363fc5b083be88002b6bR25) so I do indeed think its required. If just the environment variable is enough with `*`s pattern matching that seems fine to me.

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh (hunk 1 above)
Comment (corngood) [reply pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:55]: Not sure what you mean in your last sentence (do you mean current upstream nuget behaviour?), but unless something has changed upstream, the actual source used by nuget is non-deterministic.

There's some discussion of it here:

https://github.com/NuGet/Home/issues/3676

I believe it uses local sources first, but for the remote ones it's just whoever responds first.

You should be able to reproduce something like the example I posted at the top using the omnisharp-roslyn fetch script, since it has a bunch of sources specified.

**Comment by corngood** [comment]
This is probably off-topic, but I was actually thinking about `systemToDotnetRid`.  If you look how it's used in ryujinx:

```nix
  dotnetRestoreFlags = [ "--runtime ${dotnetCorePackages.sdk_6_0.systemToDotnetRid stdenvNoCC.targetPlatform.system}" ];
```

There's no relation to the sdk used by `buildDotnetModule`.  Maybe we should just put it in `dotnetCorePackages` instead of the sdk?

I'll do something in another PR for this.

**Comment by corngood** [comment]
I think I got everything addressed.  The sort order is much closer to before, but there's still stuff like this:

```diff
-  (fetchNuGet { pname = "Microsoft.Extensions.Logging"; version = "3.1.18"; sha256 = "0llw82p6crp0329n3rsyrqka21c3dqyjk8lbk25y5848vzi0bzbv"; })
   (fetchNuGet { pname = "Microsoft.Extensions.Logging"; version = "3.1.1"; sha256 = "1y78bn463mrl8vy7iwafrmq4x0vg4pqjd3xaiznfg9lpxjgjl9j3"; })
+  (fetchNuGet { pname = "Microsoft.Extensions.Logging"; version = "3.1.18"; sha256 = "0llw82p6crp0329n3rsyrqka21c3dqyjk8lbk25y5848vzi0bzbv"; })
   (fetchNuGet { pname = "Microsoft.Extensions.Logging.Abstractions"; version = "3.1.0"; sha256 = "1zyalrcksszmn9r5xjnirfh7847axncgzxkk3k5srbvlcch8fw8g"; })
-  (fetchNuGet { pname = "Microsoft.Extensions.Logging.Abstractions"; version = "3.1.18"; sha256 = "0pq1kw77zz9ygcdw87wxd1rkcij084jj1cgp6p4b8zpl0a73ba6b"; })
   (fetchNuGet { pname = "Microsoft.Extensions.Logging.Abstractions"; version = "3.1.1"; sha256 = "0pyk6g2qs1lrjhj1qz4bqbqpbmbgqlah1b6ynlvv5bdsrb7157zf"; })
+  (fetchNuGet { pname = "Microsoft.Extensions.Logging.Abstractions"; version = "3.1.18"; sha256 = "0pq1kw77zz9ygcdw87wxd1rkcij084jj1cgp6p4b8zpl0a73ba6b"; })
```

Because versions are now being sorted using globs instead of sorting the actual contents of the output file.  I think this way is saner though.

File: pkgs/development/dotnet-modules/python-language-server/default.nix [hunk 5]
```diff
@@ -25,6 +26,8 @@ buildDotnetModule rec {
   dotnet-sdk = dotnetCorePackages.sdk_3_1;
   dotnet-runtime = dotnetCorePackages.runtime_3_1;
 
+  dotnetRestoreFlags = [ "--runtime ${dotnet-sdk.systemToDotnetRid stdenvNoCC.targetPlatform.system}" ];
```
Comment (mdarocha) [inline pkgs/development/dotnet-modules/python-language-server/default.nix:29]: Maybe we should consider adding this flag by default to dotnet-restore-hook.sh?

If yes, then we should also consider adding an ability to override it (when ie. building a windows version)

**Review by mdarocha — APPROVED** [review]
LGTM, dont consider my comments blocking

File: pkgs/development/dotnet-modules/python-language-server/default.nix (hunk 5 above)
Comment (corngood) [reply pkgs/development/dotnet-modules/python-language-server/default.nix:29]: I actually did try that locally: https://github.com/corngood/nixpkgs/commit/204c2998506152d0c1d4a744ad1f6dbf51233c33

It breaks a bunch of packages: https://gist.github.com/corngood/902393ad085e7b2d447e52b877dbfde3

I still think it's a good idea, but I'm not sure it can be done universally due to all the hacky stuff projects do with properties.  For example:

> /nix/store/byipzqmp760qz1nj0cncvjln1zpdd6xa-dotnet-sdk-6.0.401/sdk/6.0.401/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1083: The specified RuntimeIdentifier 'linux-x64;linux-x64' is not recognized. [/build/source/osu.Game.Rulesets.Catch/osu.Game.Rulesets.Catch.csproj]

Comment (mdarocha) [reply pkgs/development/dotnet-modules/python-language-server/default.nix:29]: Yeah, the cli is a bit buggy. We can make an issue for it and make the change in the future

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh [hunk 6]
```diff
@@ -3,44 +3,80 @@
...
+        fi
+      fi
+    done
+
+    if ! ${found-false}; then
+      echo "couldn't find $package $version" >&2
+      exit 1
+    fi
+
+    if [[ "$source" != https://api.nuget.org/v3/index.json ]]; then
```
[the remark this answers is not in the document]
Comment (mdarocha) [reply pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:71]: Hmm, the custom fetcher is a good idea for a future PR

**Review by mdarocha — APPROVED** [review]

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh (hunk 1 above)
Comment (IvarWithoutBones) [reply pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:56]: > do you mean current upstream nuget behaviour?

I do, I wasn't aware of this being upstream behaviour. This is not an issue then, could you just add a comment explaining that? The log itself is a bit confusing without this context IMO.

**Comment by IvarWithoutBones** [comment]
> This is probably off-topic, but I was actually thinking about `systemToDotnetRid`. If you look how it's used in ryujinx:
> 
> ```nix
>   dotnetRestoreFlags = [ "--runtime ${dotnetCorePackages.sdk_6_0.systemToDotnetRid stdenvNoCC.targetPlatform.system}" ];
> ```
> 
> There's no relation to the sdk used by `buildDotnetModule`. Maybe we should just put it in `dotnetCorePackages` instead of the sdk?
> 
> I'll do something in another PR for this.

Sounds good to me. I wasn't sure where to put it when creating `systemToDotnetRid`, I agree the SDK isnt the best place. I think an `dotnetPlatform` attribute makes a lot of sense, much like our rust infrastructure has. That could house `mkNugetDeps`/`mkNugetSource` and maybe `nuget-to-nix` as well.

File: pkgs/development/dotnet-modules/python-language-server/default.nix (hunk 5 above)
Comment (IvarWithoutBones) [reply pkgs/development/dotnet-modules/python-language-server/default.nix:29]: Since it is such a common flag it would maybe be nice to have an `dotnetUseRid` option so that `buildDotnetModule` can set it for you.

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh [hunk 7]
```diff
@@ -3,44 +3,80 @@
...
+      jq -r '.resources[] | select(."@type" == "PackageBaseAddress/3.0.0")."@id"')
+done
 
-  if grep "$pkg_name" "$exclusions" > /dev/null; then
-    continue
-  fi
+echo "{ fetchNuGet }: ["
 
-  pkg_sha256="$(nix-hash --type sha256 --flat --base32 "$(dirname "$pkg_spec")"/*.nupkg)"
+cd "$1"
```
Comment (IvarWithoutBones) [inline pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:34]:
```suggestion
cd "$pkgs"
```

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh [hunk 8]
```diff
@@ -3,44 +3,80 @@
...
+  >&2 echo "Usage: $0 <packages directory> [path to excluded package source] > deps.nix"
   exit 1
 fi
 
 pkgs=$1
-exclusions="${2:-/dev/null}"
-tmpfile=$(mktemp /tmp/nuget-to-nix.XXXXXX)
-trap "rm -f ${tmpfile}" EXIT
+tmp=$(realpath "$(mktemp -td nuget-to-nix.XXXXXX)")
+trap "rm -r $tmp" EXIT
```
Comment (IvarWithoutBones) [inline pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh:16]: Should use single quotes so that `$tmp` gets expanded upon `EXIT` instead of immidately.
```suggestion
trap 'rm -r $tmp' EXIT
```

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh [hunk 9]
```diff
@@ -3,44 +3,80 @@
...
+    for source in "${sources[@]}"; do
+      url="${base_addresses[$source]}$package/$version/$package.$version.nupkg"
+      if [[ "$source" == "$used_source" ]]; then
+        sha256="$(nix-hash --type sha256 --flat --base32 "$version/$package.$version".nupkg)"
+        found=true
+        break
+      else
+        if sha256=$(nix-prefetch-url "$url" 2>"$tmp"/error); then
+          echo $package $version is available on $url, but was downloaded from ${base_addresses[$used_source]}$package/$version/$package.$version.nupkg 1>&2
+          # exit 1
```
Comment (IvarWithoutBones) [inline pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh]:
```suggestion
          echo "$package $version is available on $url, but was downloaded from ${base_addresses[$used_source]}$package/$version/$package.$version.nupkg" 1>&2
```

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh (hunk 1 above)
Comment (corngood) [reply pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh]: comment added

File: pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh (hunk 9 above)
Comment (IvarWithoutBones) [reply pkgs/build-support/dotnet/nuget-to-nix/nuget-to-nix.sh]: Comment is a bit outdated now but this should still get quoted

**Comment by IvarWithoutBones** [comment]
I would also prefer for the [`nuget-to-nix: find sources deterministically`](https://github.com/NixOS/nixpkgs/pull/193033/commits/ad152338af6b4161f389e3dc21f8bceb4d8e3125) commit to only contain changes to `nuget-to-nix` itself, the regenerating of lockfiles should be done seperately.

File: pkgs/build-support/dotnet/build-dotnet-module/default.nix [hunk 10]
```diff
@@ -188,18 +190,18 @@ stdenvNoCC.mkDerivation (args // {
...
             if test -n "''${keepSources-}"; then
-                echo -e "Path to the source: $src\nPath to the fake home: $HOME"
+                echo -e "Path to the source: $tmp/src\nPath to the fake home: $tmp/home"
             else
-                rm -rf "$src" "$HOME"
+                rm -rf "$tmp"
             fi
-
-            # Since mktemp is used this will be empty if the script didnt succesfully complete
-            ! test -s "$depsFile" && rm -rf "$depsFile"
```
Comment (IvarWithoutBones) [inline pkgs/build-support/dotnet/build-dotnet-module/default.nix:202]: Why was this removed?

Comment (corngood) [reply pkgs/build-support/dotnet/build-dotnet-module/default.nix]: That was a mistake, at one point I had moved the file into $tmp, but put it back and forgot to undo this change.

**Review by IvarWithoutBones — APPROVED** [review]
Regenerated lockfiles of a few derivations and confirmed they're consistent on darwin. This looks good to me, thanks!

Result of `nixpkgs-review pr 193033` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>baget</li>
    <li>depotdownloader</li>
    <li>github-runner</li>
    <li>jellyfin</li>
    <li>nuget-to-nix</li>
    <li>omnisharp-roslyn</li>
    <li>opentabletdriver</li>
    <li>osu-lazer</li>
    <li>pinta</li>
    <li>python-language-server</li>
    <li>ryujinx</li>
    <li>xivlauncher</li>
  </ul>
</details>
